{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEPENDENCIES\n",
    "\n",
    "import pysam\n",
    "import pandas as pd\n",
    "import HTSeq\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "import matplotlib.pyplot as plt\n",
    "# from tqdm.notebook import tqdm, trange\n",
    "import pstats\n",
    "from tqdm import tqdm as tqdm\n",
    "import cProfile\n",
    "import plotly.io as pio\n",
    "pio.renderers.default = 'notebook_connected'\n",
    "import numpy as np\n",
    "pd.set_option('display.max_rows', 100)\n",
    "import chart_studio.tools as tls\n",
    "\n",
    "# GLOBAL VARS\n",
    "HTML_FOLDER =\"misc/interactive_figures/\"\n",
    "FC30_DMGOTH_MAX_AS_PRIMARY_ONLY_BAMFILE =\"data/dmgoth101_genome_alignments/bam/FC30.against_dmgoth.filtered_max_AS.primary_only.bam\"\n",
    "FC29_DMGOTH_MAX_AS_PRIMARY_ONLY_BAMFILE =\"data/dmgoth101_genome_alignments/bam/FC29.against_dmgoth.filtered_max_AS.primary_only.bam\"\n",
    "GENE_ANNOTATIONS_FILE = \"/data/dmgoth101_genome_alignments/annotations/Dm_Goth_10-1.dmel6.23LiftOff.sorted.gff\"\n",
    "\n",
    "# TE_CLASSIFICATION_FILE= \"/data2/eric/TE_LR_RNAseq/TE_hierarchy.tsv\"\n",
    "TE_CLASSIFICATION_FILE_V3= \"data/dmgoth101_genome_alignments/annotations/TE_classification.from_RM_output_v3.csv\"\n",
    "TE_CLASSIFICATION_FILE_V2= \"data/dmgoth101_genome_alignments/annotations/TE_classification.from_RM_output_v2.csv\"\n",
    "# BLAST_TE_ANNOTATIONS_FILE = \"/data2/eric/TE_LR_RNAseq/data/dmgoth101_genome_alignments/annotations/Dm_Goth_10-1_insertions_vsTEdb.gtf\"\n",
    "RM_TE_ANNOTATION_FILE_V3 = \"data/dmgoth101_genome_alignments/annotations/dmgoth101.onecode.v3.gtf\"\n",
    "RM_TE_ANNOTATION_FILE_V2 = \"data/dmgoth101_genome_alignments/annotations/dmgoth101.onecode.v2.gtf\"\n",
    "\n",
    "# INSERTION_TABLE = \"/data2/eric/TE_LR_RNAseq/data/dmgoth101_genome_alignments/insertion_table.tsv\"\n",
    "\n",
    "# GLOBAL PARAMETERS\n",
    "\n",
    "MIN_SUB_COVERAGE = 0.1 # Threshold used to filter the features (gene or TE) mapped by a read = minimal subject coverage (nb of aligned bases / total nb of feature's bases)\n",
    "\n",
    "PLOTLY_SHOW_CONFIG = {\n",
    "  'toImageButtonOptions': {\n",
    "    'format': 'svg', # one of png, svg, jpeg, webp\n",
    "    'filename': 'custom_image',\n",
    "    'height': None,\n",
    "    'width': None,\n",
    "    'scale': 1 # Multiply title/legend/axis/canvas sizes by this factor\n",
    "  }\n",
    "}\n",
    "\n",
    "import chart_studio\n",
    "import chart_studio.plotly as py\n",
    "\n",
    "chart_studio.tools.set_credentials_file(username='EricCumunel', api_key='ve7YItpOOE4o6GRAWxYv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saved_FC29_counting_df_v3 = pd.read_csv(\"countings/FC29_counting_df.v3.tsv\", sep = \"\\t\")\n",
    "filtered_FC29_counting_df_v3 = saved_FC29_counting_df_v3[saved_FC29_counting_df_v3['Counting'] >= 6]\n",
    "\n",
    "saved_FC30_counting_df_v3 = pd.read_csv(\"countings/FC30_counting_df.v3.tsv\", sep = \"\\t\")\n",
    "filtered_FC30_counting_df_v3 = saved_FC30_counting_df_v3[saved_FC30_counting_df_v3['Counting'] >= 2]\n",
    "\n",
    "FC29_counters = [22232, 3913, 1524, 1476, 1306, 5126, 8887]\n",
    "FC30_counters = [7332, 489, 101, 452, 280, 1386, 4624]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Male funnel plot\n",
    "def draw_funnel_plot(counters, title):\n",
    "    tot = counters[0]\n",
    "    filter1 = tot - counters[5]\n",
    "    filter2 = filter1 - counters[6]\n",
    "\n",
    "    data = dict(\n",
    "        number=[tot, filter1, filter2 ],\n",
    "        Filters=[\"Reads aligned on a TE\", \"Reads with at least 10% subject cover\", \"Reads with less non-overlapping bases on TE\"])\n",
    "    fig = px.funnel(data, x='number', y='Filters', title=title)\n",
    "    fig.show(config=PLOTLY_SHOW_CONFIG)\n",
    "    # fig.write_html(HTML_FOLDER + \"male_funnel_plot.html\")\n",
    "\n",
    "draw_funnel_plot(FC29_counters, \"Male Dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_funnel_plot(FC30_counters, \"Female dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_counters_pie_chart(counters_1, counters_2):\n",
    "\tlabels = ['Aligned on one TE only','Aligned on multiple TE','Aligned in exonic region','Aligned in intronic region']\n",
    "\n",
    "\t# Create subplots: use 'domain' type for Pie subplot\n",
    "\tfig = make_subplots(rows=1, cols=2, specs=[[{'type':'domain'}, {'type':'domain'}]])\n",
    "\tfig.add_trace(go.Pie(labels=labels, values=counters_1[1:], name=\"Male dataset\"), 1, 1)\n",
    "\tfig.add_trace(go.Pie(labels=labels, values=counters_2[1:], name=\"Female dataset\"), 1, 2)\n",
    "\n",
    "\t# Use `hole` to create a donut-like pie chart\n",
    "\tfig.update_traces(hole=.3, textinfo='value', hoverinfo=\"label+value+name\")\n",
    "\n",
    "\tfig.update_layout(\n",
    "\t\ttitle_text=\"Context of reads mapped on TE\",\n",
    "\t\t# Add annotations in the center of the donut pies.\n",
    "\t\tannotations=[dict(text='Male', x=0.20, y=0.5, font_size=20, showarrow=False),\n",
    "\t\t\t\t\tdict(text='Female', x=0.81, y=0.5, font_size=20, showarrow=False)])\n",
    "\tfig.show(config=PLOTLY_SHOW_CONFIG)\n",
    "\n",
    "draw_counters_pie_chart(FC29_counters, FC30_counters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_icicle(df):\n",
    "    fig = px.icicle(df, path=['Subclass', 'Superfamily', 'Family', 'Insertion'], values='Counting',\n",
    "                    color='mean_subcov',\n",
    "                    hover_data=['Counting'],\n",
    "                    color_continuous_scale='RdBu',\n",
    "                    color_continuous_midpoint=np.average(df['mean_subcov'],weights=df['Counting'])\n",
    "                    )\n",
    "    fig.update_traces(root_color=\"lightgrey\")\n",
    "    fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))\n",
    "    fig.show()\n",
    "\n",
    "def draw_sunburst(df):\n",
    "    fig = px.sunburst(df, path=['Subclass', 'Superfamily', 'Family', 'Insertion'], values='Counting',\n",
    "                    color='mean_subcov',\n",
    "                    hover_data=['Counting'],\n",
    "                    color_continuous_scale='Plasma',\n",
    "                    # color_continuous_scale='RdBu',\n",
    "                    # color_continuous_midpoint=np.average(df['mean_subcov'],weights=df['Counting'])\n",
    "                    color_continuous_midpoint=np.average(0.5)\n",
    "                    )\n",
    "    fig.update_traces(root_color=\"lightgrey\")\n",
    "    fig.update_layout(margin = dict(t=50, l=25, r=25, b=25))\n",
    "    fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_2_sunburst_charts(counting_1, counting_2):\n",
    "    fig = make_subplots(rows=1, cols=2, specs=[[{\"type\": \"sunburst\"}, {\"type\": \"sunburst\"}]])\n",
    "    def make_sunburst(counting):\n",
    "        return px.sunburst(counting, path=['Subclass', 'Superfamily', 'Family', 'Insertion'],\n",
    "        values='Counting',\n",
    "        color='mean_subcov',\n",
    "        hover_data=['Counting'],\n",
    "        color_continuous_scale='Plasma',\n",
    "        color_continuous_midpoint=np.average(counting['mean_subcov'], weights=counting['mean_subcov'])\n",
    "        )\n",
    "    sunburst1 = make_sunburst(counting_1)\n",
    "    sunburst2 = make_sunburst(counting_2)\n",
    "    fig.add_trace(sunburst1.data[0], row=1, col=1)\n",
    "    fig.add_trace(sunburst2.data[0], row=1, col=2)\n",
    "    fig.show(config=PLOTLY_SHOW_CONFIG)\n",
    "    fig.write_html(HTML_FOLDER + \"TE_LR_RNAseq_sunbursts_comparison.html\")\n",
    "    return fig\n",
    "\n",
    "sunburst = draw_2_sunburst_charts(filtered_FC29_counting_df_v3, filtered_FC30_counting_df_v3)\n",
    "# py.plot(sunburst, filename = 'TE_LR_RNAseq_sunbursts', auto_open=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "family_list = ['POGO', 'Copia_LTR', 'ROO_I', 'Gypsy12_LTR', 'DNAREP1_DM', 'Mariner2_DM', \"BARI_DM\"]\n",
    "# family_list = ['POGO', 'Copia_LTR', 'ROO_I', 'Gypsy12_LTR', 'DNAREP1_DM', 'Mariner2_DM']\n",
    "\n",
    "def get_family_list(counting_df, subclass):\n",
    "\tfamily_list = list(set(counting_df[(counting_df[\"Counting\"]) & (counting_df[\"Subclass\"] == subclass)][\"Family\"]))\n",
    "\treturn family_list\n",
    "\n",
    "# family_list = ['MINOS']\n",
    "def draw_violin_charts(counting_df_1, counting_df_2, family_list):\n",
    "\tcolor_list = ['#636EFA', '#EF553B', '#00CC96', '#AB63FA', '#FFA15A', '#19D3F3', '#FF6692', '#B6E880', '#FF97FF', '#FECB52']\n",
    "\twhile len(family_list) > len(color_list):\n",
    "\t\tcolor_list += color_list\n",
    "\tfig = make_subplots(rows=2, cols=1,\n",
    "\tshared_xaxes=True,\n",
    "\tshared_yaxes='all',\n",
    "\trow_titles=[\"Male\", \"Female\"],\n",
    "\t)\n",
    "\n",
    "\tdef make_violin(counting_df, family, color, show_legend):\n",
    "\t\ttot_nb_of_read = sum(counting_df[\"Counting\"])\n",
    "\t\tdf = counting_df[(counting_df[\"Family\"] == family) & (counting_df[\"Counting\"] > 0)]\n",
    "\t\treturn go.Violin(x=df[\"Family\"],\n",
    "\t\t\t\ty=np.log10(df['Counting']/tot_nb_of_read),\n",
    "\t\t\t\t# y=df[\"Counting\"],\n",
    "\t\t\t\tname=family,\n",
    "\t\t\t\tpoints=\"all\",\n",
    "\t\t\t\tpointpos=0,\n",
    "\t\t\t\t# box_visible=True,\n",
    "\t\t\t\tmeanline_visible=True,\n",
    "\t\t\t\tline_color=\"black\",\n",
    "\t\t\t\topacity = 0.9,\n",
    "\t\t\t\tcustomdata = np.stack((df['Insertion'], df['Counting']), axis=-1),\n",
    "\t\t\t\thovertemplate = ('<b>Insertion</b>: %{customdata[0]}<br>'+'<b>Counting</b>: %{customdata[1]}'),\n",
    "\t\t\t\tlegendgroup=family,\n",
    "\t\t\t\tshowlegend = show_legend,\n",
    "\t\t\t\tfillcolor=color)\n",
    "\n",
    "\n",
    "\tfor row, df in enumerate([counting_df_1, counting_df_2]):\n",
    "\t\tfor col, family in enumerate(family_list):\n",
    "\t\t\tfig.add_trace(make_violin(df, family, color_list[col], bool(row)), row=row+1, col=1)\n",
    "\tfig.update_layout()\n",
    "\tfig.show(config=PLOTLY_SHOW_CONFIG)\n",
    "\treturn fig\n",
    "\n",
    "fig = draw_violin_charts(saved_FC29_counting_df_v3, saved_FC30_counting_df_v3, family_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_all_violin_plots(male_counting_df, female_counting_df):\n",
    "\tsubclass_set = set(list(male_counting_df[\"Subclass\"].unique()) +  list(female_counting_df[\"Subclass\"].unique()))\n",
    "\tfor subclass in subclass_set:\n",
    "\t\tprint(subclass)\n",
    "\t\tfamily_list = get_family_list(male_counting_df, subclass)\n",
    "\t\tfamily_list.extend(get_family_list(female_counting_df, subclass))\n",
    "\t\tfamily_list = list(set(family_list))\n",
    "\t\tnew_fig = draw_violin_charts(male_counting_df, female_counting_df, family_list)\n",
    "\t\t# new_fig.write_html(HTML_FOLDER + \"TE_LR_RNAseq_violin_plot.{}.log2.html\".format(subclass))\n",
    "\t\t# py.plot(new_fig, filename = 'TE_LR_RNAseq_violin.{}.log2'.format(subclass), auto_open=False)\n",
    "generate_all_violin_plots(saved_FC29_counting_df_v3, saved_FC30_counting_df_v3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_insertion_length(insertion_name):\n",
    "\tstart, end = insertion_name.split('$')[-2:]\n",
    "\treturn int(end) - int(start) + 1\n",
    "\n",
    "def get_insertion_merged_df(female_counting, male_counting):\n",
    "\tinsertion_list = set(list(female_counting[\"Insertion\"]) + list(male_counting[\"Insertion\"]))\n",
    "\tsubclass_list = []\n",
    "\tsuperfamily_list = []\n",
    "\tfamily_list = []\n",
    "\tmale_counting_list = []\n",
    "\tfemale_counting_list = []\n",
    "\tinsertion_length_list = []\n",
    "\tfor insertion in insertion_list :\n",
    "\t\tif insertion in list(female_counting[\"Insertion\"]) and insertion in list(male_counting[\"Insertion\"]):\n",
    "\t\t\tinsertion_df = female_counting[female_counting[\"Insertion\"] == insertion]\n",
    "\t\t\tfemale_counting_list.append(insertion_df[\"Counting\"].values[0])\n",
    "\t\t\tmale_counting_list.append(male_counting[male_counting[\"Insertion\"] == insertion][\"Counting\"].values[0])\n",
    "\t\t\t\n",
    "\t\telif insertion in list(female_counting[\"Insertion\"]) :\n",
    "\t\t\tinsertion_df = female_counting[female_counting[\"Insertion\"] == insertion]\n",
    "\t\t\tfemale_counting_list.append(insertion_df[\"Counting\"].values[0])\n",
    "\t\t\tmale_counting_list.append(0)\n",
    "\t\telse :\n",
    "\t\t\tinsertion_df = male_counting[male_counting[\"Insertion\"] == insertion]\n",
    "\t\t\tmale_counting_list.append(insertion_df[\"Counting\"].values[0])\n",
    "\t\t\tfemale_counting_list.append(0)\n",
    "\t\tsubclass_list.append(insertion_df[\"Subclass\"].values[0])\n",
    "\t\tsuperfamily_list.append(insertion_df[\"Superfamily\"].values[0])\n",
    "\t\tfamily_list.append(insertion_df[\"Family\"].values[0])\n",
    "\t\tinsertion_length_list.append(get_insertion_length(insertion))\n",
    "\n",
    "\tinsertion_merged_df = pd.DataFrame(list(zip(subclass_list, superfamily_list, family_list, insertion_list, insertion_length_list, female_counting_list, male_counting_list )), columns=[\"Subclass\", \"Superfamily\", \"Family\", \"Insertion\", \"Length\", \"Female_counting\", \"Male_counting\"])\n",
    "\n",
    "\treturn insertion_merged_df\n",
    "\n",
    "insertion_merged_df = get_insertion_merged_df(saved_FC30_counting_df_v3, saved_FC29_counting_df_v3)\n",
    "# print(insertion_merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = insertion_merged_df\n",
    "fig = px.scatter(df, x=\"Female_counting\", y=\"Male_counting\", color=\"Subclass\",\n",
    "                #  size='petal_length',\n",
    "\t\t\t\t hover_data=['Subclass', 'Superfamily', 'Family', 'Insertion', 'Female_counting', 'Male_counting', 'Length'],\n",
    "\t\t\t\t#  trendline=\"ols\"\n",
    "\t\t\t\t )\n",
    "fig.update_xaxes(range=[0, 100])\n",
    "fig.update_yaxes(range=[0, 200])\n",
    "fig.show(config=PLOTLY_SHOW_CONFIG)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
